src/workerd/api/node/zlib-util.h

```diff
@@ -1,57 +1,261 @@
 // Copyright (c) 2017-2022 Cloudflare, Inc.
 // Licensed under the Apache 2.0 license found in the LICENSE file or at:
 //     https://opensource.org/licenses/Apache-2.0
-
+// Copyright Joyent and Node contributors. All rights reserved. MIT license.
 #include <workerd/jsg/jsg.h>
+#include <workerd/jsg/exception.h>
 
+#include "zlib.h"
 #include <kj/array.h>
 #include <kj/compat/brotli.h>
-#include "zlib.h"
+#include <kj/mutex.h>
+#include <kj/vector.h>
 
-#include <limits>
-#include <cstdint>
+#include <cstdlib>
 
 namespace workerd::api::node {
 
+#ifndef ZLIB_ERROR_CODES
+#define ZLIB_ERROR_CODES(V)      \
+  V(Z_OK)                        \
+  V(Z_STREAM_END)                \
+  V(Z_NEED_DICT)                 \
+  V(Z_ERRNO)                     \
+  V(Z_STREAM_ERROR)              \
+  V(Z_DATA_ERROR)                \
+  V(Z_MEM_ERROR)                 \
+  V(Z_BUF_ERROR)                 \
+  V(Z_VERSION_ERROR)             \
+
+inline const char* ZlibStrerror(int err) {
+#define V(code) if (err == code) return #code;
+  ZLIB_ERROR_CODES(V)
+#undef V
+  return "Z_UNKNOWN_ERROR";
+}
+#endif  // ZLIB_ERROR_CODES
+
+// Certain zlib constants are defined by Node.js itself
+static constexpr auto Z_MIN_CHUNK = 64;
+static constexpr auto Z_MAX_CHUNK = 128 * 1024;
+static constexpr auto Z_DEFAULT_CHUNK = 16 * 1024;
+static constexpr auto Z_MIN_MEMLEVEL = 1;
+
+static constexpr auto Z_MAX_MEMLEVEL = 9;
+static constexpr auto Z_DEFAULT_MEMLEVEL = 8;
+static constexpr auto Z_MIN_LEVEL = -1;
+static constexpr auto Z_MAX_LEVEL = 9;
+static constexpr auto Z_DEFAULT_LEVEL = Z_DEFAULT_COMPRESSION;
+static constexpr auto Z_MIN_WINDOWBITS = 8;
+static constexpr auto Z_MAX_WINDOWBITS = 15;
+static constexpr auto Z_DEFAULT_WINDOWBITS = 15;
+
+static constexpr uint8_t GZIP_HEADER_ID1 = 0x1f;
+static constexpr uint8_t GZIP_HEADER_ID2 = 0x8b;
+
+using ZlibModeValue = uint8_t;
+enum class ZlibMode : ZlibModeValue {
+  NONE,
+  DEFLATE,
+  INFLATE,
+  GZIP,
+  GUNZIP,
+  DEFLATERAW,
+  INFLATERAW,
+  UNZIP,
+  BROTLI_DECODE,
+  BROTLI_ENCODE
+};
+
+struct CompressionError {
+  CompressionError(kj::StringPtr _message, kj::StringPtr _code, int _err)
+      : message(kj::str(_message)),
+        code(kj::str(_code)),
+        err(_err) {
+    JSG_REQUIRE(message.size() != 0, Error, "Compression error message should not be null");
+  }
+
+  kj::String message;
+  kj::String code;
+  int err;
+};
+
+class ZlibContext final {
+public:
+  ZlibContext() = default;
+
+  KJ_DISALLOW_COPY(ZlibContext);
+
+  void close();
+  void setBuffers(kj::ArrayPtr<kj::byte> input, uint32_t inputLength, kj::ArrayPtr<kj::byte> output, uint32_t outputLength) {
```

Comment:
What is our convention on when it's ok to implement methods inline? My instinct would be to avoid it, because it slows down compilation time, but this is a project-dependent thing.

Reply:
i'm not sure. i have this way in my mind that if the function is simple to implement in header, i put it in header. but technically we should have `zlib-util-inl.h` class if we want to conform to Google code style. @jasnell?

Reply:
There may be some difference of opinion on it in the team but for me, it really comes down to the complexity of the function. Generally try to avoid inlining in the headers unless the body is very trivial (e.g. a trivial constructor, or a trivial getter/setter). When in doubt, if it's going to be more than a few lines, don't inline.

Reply:
@jasnell That's my usual approach as well. However wouldn't we get faster builds if we totally crack down on inline implementations?